src/Widget/Conversation/Messages/Messages.tsx

```diff
@@ -17,40 +18,81 @@ function Messages() {
   const { appState } = useAppContext();
   const { animations } = appState;
 
-  const messageRef = useRef<HTMLUListElement | null>(null);
+  const containerRef = useRef<HTMLUListElement | null>(null);
 
   const prevMessagesHistory = usePrevious(appState.messagesHistory);
   const widgetHeight = useWidgetHeight();
 
+  const [messagessIdsInViewPoint, setMessagesIdsInViewpoint] = useState<string[]>([]);
+
   useEffect(() => {
     if (prevMessagesHistory) {
       const newMessageIsCreating = !isEqual(prevMessagesHistory, appState.messagesHistory);
 
-      scrollToBottom(messageRef.current, newMessageIsCreating);
+      scrollToBottom(containerRef.current, newMessageIsCreating);
       return;
     }
 
-    scrollToBottom(messageRef.current);
+    scrollToBottom(containerRef.current);
   }, [appState, prevMessagesHistory]);
 
   const inlineHeightStyles = window.innerWidth > 600 ? { maxHeight: widgetHeight } : {};
 
+  const manageAnimationsVisibility = throttle(() => {
+    if (!containerRef.current) return;
+    const { current } = containerRef;
+
+    const { top: containerTop, bottom: containerBottom } = current.getBoundingClientRect();
+
+    const elmentsWithAnimations = document.querySelectorAll('[data-msg-with-animation]');
+
+    Array.from(elmentsWithAnimations).forEach((el: HTMLElement) => {
+      const { bottom, top } = el.getBoundingClientRect();
+
+      const isVisible = !(bottom < containerTop || top > containerBottom);
+      const messageId = el.dataset.msgHistoryId as string;
+
+      if (isVisible) {
```

Comment:
Why do we need to repeatedly patch `messagessIdsInViewPoint`? Can we do something like:
```
Array.from(elmentsWithAnimations).filter( ... visibility stuff ... ).map((el: HTMLElement) => el.dataset.msgHistoryId as string);
setMessagesIdsInViewpoint(messagessIdsInViewPoint);
```
After all we just want ids of visible elements and we can get them at once instead of iteratively removing/adding them to mold messagessIdsInViewPoint in many passes.

Unless I am not seeing some important clever wrinkle here.